File: src/java/org/voegtle/weatherstation/server/request/AbstractUrlParameter.java
package org.voegtle.weatherstation.server.request;

import org.voegtle.weatherstation.server.util.DateUtil;
import org.voegtle.weatherstation.server.util.StringUtil;

import javax.servlet.http.HttpServletRequest;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

abstract class AbstractUrlParameter {
  private static final String FORMAT_TIMESTAMP = "yyyy-MM-dd-HH:mm:ss";

  final private HttpServletRequest request;
  private final DateUtil dateUtil;

  protected AbstractUrlParameter(final HttpServletRequest request, DateUtil dateUtil) {
    this.request = request;
    this.dateUtil = dateUtil;
  }

  protected String getUrlParameter(String paramName) {
    String param = request.getParameter(paramName);
    if ("".equals(param)) {
      return null;
    }
    return param;
  }

  protected boolean getUrlParameterBoolean(String paramName) {
    String param = request.getParameter(paramName);
    return param != null;
  }

  protected Date getUrlParameterDate(String paramName, boolean localTimezone) {
    String param = request.getParameter(paramName);
    if (StringUtil.isEmpty(param)) {
      return null;
    }

    SimpleDateFormat sdf = new SimpleDateFormat(FORMAT_TIMESTAMP);
    Date result;
    try {
      result = sdf.parse(param);
      if (localTimezone) {
        result = dateUtil.fromLocalToGMT(result);
        result = dateUtil.fromGMTtoCEST(result);
      } else {
        result = dateUtil.fromCESTtoGMT(result);
      }
    } catch (ParseException e) {
      result = null;
    }

    return result;
  }

  protected Integer getUrlParameterInteger(String paramName) {
    return getUrlParameterInteger(paramName, null);
  }

  protected Integer getUrlParameterInteger(String paramName, Integer defaultValue) {
    String param = request.getParameter(paramName);
    if (StringUtil.isNotEmpty(param)) {
      return new Integer(param);
    }
    return defaultValue;
  }

  protected DataType getUrlParameterType(String paramName, DataType defaultType) {
    String param = request.getParameter(paramName);
    DataType dataType = DataType.fromString(param);
    if (dataType.equals(DataType.UNDEFINED)) {
      dataType = defaultType;
    }
    return dataType;
  }

  protected DataType getUrlParameterType(String paramName) {
    String param = request.getParameter(paramName);
    return DataType.fromString(param);
  }
}

File: src/main/java/org/voegtle/weatherstation/server/request/AbstractUrlParameter.kt
package org.voegtle.weatherstation.server.request

import org.voegtle.weatherstation.server.util.DateUtil
import org.voegtle.weatherstation.server.util.StringUtil
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Date
import javax.servlet.http.HttpServletRequest

abstract class AbstractUrlParameter protected constructor(private val request: HttpServletRequest,
                                                               private val dateUtil: DateUtil) {
  private val FORMAT_TIMESTAMP = "yyyy-MM-dd-HH:mm:ss"

  protected fun getUrlParameter(paramName: String): String? {
    val param = request.getParameter(paramName)
    return if ("" == param) null else param
  }

  protected fun getUrlParameterBoolean(paramName: String): Boolean {
    val param = request.getParameter(paramName)
    return param != null
  }

  protected fun getUrlParameterDate(paramName: String, localTimezone: Boolean): Date? {
    val param = request.getParameter(paramName)
    if (StringUtil.isEmpty(param)) {
      return null
    }

    val sdf = SimpleDateFormat(FORMAT_TIMESTAMP)
    var result: Date?
    try {
      result = sdf.parse(param)
      result = if (localTimezone) {
        dateUtil.fromLocalToGMT(result)
      } else {
        dateUtil.fromCESTtoGMT(result)
      }
    } catch (e: ParseException) {
      result = null
    }

    return result
  }

  @JvmOverloads protected fun getUrlParameterInteger(paramName: String, defaultValue: Int? = null): Int? {
    val param = request.getParameter(paramName)
    return if (StringUtil.isNotEmpty(param)) param.toInt() else defaultValue
  }

  protected fun getUrlParameterType(paramName: String, defaultType: DataType): DataType {
    val param = request.getParameter(paramName)
    var dataType = DataType.fromString(param)
    if (dataType == DataType.UNDEFINED) {
      dataType = defaultType
    }
    return dataType
  }

  protected fun getUrlParameterType(paramName: String): DataType {
    val param = request.getParameter(paramName)
    return DataType.fromString(param)
  }

}
